### PR TITLE
Feature/forgotpassword/57

### LINKED ISSUE #57: forgotPassword
create a forgot password

Estimate 4 hours

### PR DESCRIPTION
**Pull request by JessyBros**
User forgot his password and need to create a new one.

-Use token
-Create forgotten password for send a mail to reset password (genrate token)
-Create reset password to create a new password (check token)

### CODE REVIEW

File: src/Controller/ForgottenPasswordController.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+        //Si le formulaire est valide
+        if($form->isSubmitted() && $form->isValid()){
+
+            // On cherche si un utilisateur a cet email
+            $user = $userRepo->findOneByEmail($form->getData()['email']);
+
+            // Si l'utiliateur n'existe pas
+            if(!$user){
+                $this->addFlash('danger', 'Cette adresse n\'existe pas');
```
Comment (crazymad) [inline src/Controller/ForgottenPasswordController.php]: try to not let a user know about the existence of an address in your database, but that if the email was found he should receive an email

File: src/Controller/ForgottenPasswordController.php [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                $entityManager = $this->getDoctrine()->getManager();
+                $entityManager->persist($user);
+                $entityManager->flush();
+            }catch(\Exception $e){
+                $this->addFlash('warning', 'Une erreur est survenue : ' . $e->getMessage());
+                $this->redirectToRoute('sign_in');
+            }
+
+            // On génère l'URL de réinitialisation de mot de passe
+            $url = $this->generateUrl('reset_password',['token' => $token], UrlGeneratorInterface::ABSOLUTE_URL);
```
Comment (crazymad) [inline src/Controller/ForgottenPasswordController.php]: handle the mailing part in a private method or a service. Your controller starts to be too long

File: src/Controller/ForgottenPasswordController.php [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+        // Si le formulaire est envoyé en méthode POST
+        if($request->isMethod('POST')){
+            // On supprime le token
+            $user->setResetToken(null);
+
+            // On chiffre le mot de passe
+            $user->setPassword($passwordEncoder->encodePassword($user, $request->request->get('password')));
+            $entityManager = $this->getDoctrine()->getManager();
+            $entityManager->persist($user);
```
Comment (crazymad) [inline src/Controller/ForgottenPasswordController.php]: no need to persist when you don't add a new entity in the ebb

File: src/Entity/User.php [hunk 4]
```diff
@@ -62,6 +62,11 @@ class User implements UserInterface
      */
     private $picture;
 
+    /**
+     * @ORM\Column(type="string", length=50, nullable=true)
+     */
+    private $reset_token;
```
Comment (crazymad) [inline src/Entity/User.php]: use camelCase in php(PSR 1, 2, 12) snake_case would be only for the database

**Review by crazymad — CHANGES_REQUESTED** [review]

**Comment by JessyBros** [comment]
I made one commit  :

- Resolve conflict

- update add flash for user who ask forgotpassword. They must not know if the mail exist already or not, just a message 'mail send if this email exist'

- Create a mail service to send a mail 'forgot password'

- Update property in entity User, reset_token  to resetToken.

- remove persist to add token in database ( no need)

**Review by crazymad — APPROVED** [review]

**Pull request merged**